### PR TITLE
Prune defaults from submitted data

### LINKED ISSUE #24: Don't render parameters which have default value?
For the global parameters if you filled the required fields (molecules, run_dir) then the toml has following parameters:

```yaml
concat = 1
mode = 'local'
ncores = 8
queue = ''
queue_limit = 100
self_contained = false
```

These all have the default values specified at https://github.com/haddocking/haddock3/blob/main/src/haddock/modules/defaults.yaml
Including them makes the file look very complicated, but also very explicit, so any changes to defaults will not change the reproducibility of the workflow.

Should the workflow config file exclude parameters which are set to the default?

### PR DESCRIPTION
**Pull request by sverhoeven**
Fixes #24 

To see the defaults being pruned 
1. Goto https://deploy-preview-46--wonderful-noether-53a9e8.netlify.app/
2. Click `Global parameters` button.
3. Expand execution sub form
4. Switch from Visual to Text mode
5. Fill required fields
6. Press submit
7. In the toml text representation you should not see the `cores` parameter.
9. Change the number of cores and submit then `cores` parameter should show up again.

You can do similar things for the nodes.

### CODE REVIEW

**Comment by Peter9192** [comment]
> 6\. Press submit

Can't submit because submitting the global config requires an uploaded molecules file, and I don't have one. Getting the following error: `.molecules[0] should be string`

File: src/pruner.test.ts [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+          }
+        }
+      },
+      additionalProperties: false
+    }
+
+    const result = pruneDefaults(parameters, schema)
+
+    const expected = {
+      prop1: [{ prop2: 'somethingelse' }]
```
Comment (Peter9192) [inline src/pruner.test.ts:405]: Not sure I understand this test case. When would you ever get to see a duplicate prop? 

The current test assumes the second one overwrites the first, but I don't think this is even possible through the workflow builder, is it? 

The alternative I can think of is that, say `prop1` is a function that takes `prop2` as an input. Now, if you specify multiple values for `prop2`, a reasonable expectation could be that the function will be executed for all instances of prop2. In that case, you want to conserve the length of the array.

File: src/pruner.test.ts [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+          }
+        }
+      },
+      additionalProperties: false
+    }
+
+    const result = pruneDefaults(parameters, schema)
+
+    const expected = {
+    }
```
Comment (Peter9192) [inline src/pruner.test.ts:515]:
```suggestion
    const expected = {}
```

File: src/pruner.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+              }
+            }
+            return pruneDefaults({ a: v2 }, schemaOfItemAsObject).a
+          }).filter(v2 => v2 !== undefined)
+          if (prunedV.length > 0) {
+            newParameters[k] = prunedV
+          }
+        }
+        // TODO handle defaults in prefixItems, see https://json-schema.org/understanding-json-schema/reference/array.html#id7
+        // but do not have catalog which uses tuples yet, so can delay it
```
Comment (Peter9192) [inline src/pruner.ts:38]: Let's say we encounter a form at some point that uses prefixItems, what happens now? Does it silently pass or trigger a warning/error?

**Review by Peter9192 — COMMENTED** [review]
Hey @sverhoeven nice recursive function :-)

I noticed that a couple of test functions require a change in the length of array/object props. I don't really understand why this behavior makes sense. I can think of cases where the length is in itself a significant thing, even if the values are all defaults. But perhaps that is not applicable to our use case?

**Comment by sverhoeven** [comment]
> > 6. Press submit
> 
> Can't submit because submitting the global config requires an uploaded molecules file, and I don't have one. Getting the following error: `.molecules[0] should be string`

You can download a pdb file from https://www.rcsb.org/ for example https://files.rcsb.org/download/5T89.pdb

**Comment by sverhoeven** [comment]
For `f(g(h(x),y),z)` in store.ts 

Yep it gets cryptic after double nest. 
I dont want the functions to know about each other so chaining is not an option. 
I moved it to a helper method which calls the functions one line at a time.

File: src/pruner.test.ts [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+      const result = pruneDefaults(parameters, schema)
+
+      const expected = {}
+      expect(result).toEqual(expected)
+    })
+  })
+
+  describe('given an empty string and is set to default', () => {
+    it('should not include prop', () => {
+      const parameters = {
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply src/pruner.test.ts:89]: You have found a corner case I had not tested. Added test which keeps the empty string.

File: src/pruner.test.ts (hunk 2 above)
Comment (sverhoeven) [reply src/pruner.test.ts:564]: thanks

File: src/pruner.test.ts (hunk 1 above)
Comment (sverhoeven) [reply src/pruner.test.ts:454]: There are parameters where the order/length of the array does not matter, because the index of the item is not linked to another like the index of a molecule. So the current code works for those parameters, but you are correct for parameters where the index matters it breaks. 

I do not have a way to tell if the index matters so it is safer to keep default array items to not change the indices after it.
I tried to fix it but ran into downstream problems so spinned off #55

**Review by Peter9192 — APPROVED** [review]
Thanks for the changes.
